src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -101,12 +101,199 @@ struct extent_to_write_t {
+
+  extent_to_remap_t(
+    LBAPinRef &&pin, extent_len_t new_offset,
+      extent_len_t new_len, std::optional<bufferlist> bl)
+    : type(type_t::OVERWRITE), pin(std::move(pin)),
+      new_offset(new_offset), new_len(new_len), bl(bl) {}
+};
+using extent_to_remap_list_t = std::list<extent_to_remap_t>;
+
+// Encapsulates extents to be written out using do_insertions.
+struct extent_to_insert_t {
+  enum class type_t {
+    DATA,
+    ZERO
+  };
+  type_t type;
+  /// laddr of new extent
+  laddr_t addr;
+  /// length of new extent
+  extent_len_t len;
+  /// non-nullopt if type == DATA
+  std::optional<bufferlist> bl;
+
+  extent_to_insert_t(const extent_to_insert_t &) = default;
+  extent_to_insert_t(extent_to_insert_t &&) = default;
+
+  bool is_data() const {
+    return type == type_t::DATA;
+  }
+
+  bool is_zero() const {
+    return type == type_t::ZERO;
+  }
+
+  static extent_to_insert_t create_data(
+    laddr_t addr, extent_len_t len, std::optional<bufferlist> bl) {
+    return extent_to_insert_t(addr, len, bl);
+  }
+
+  static extent_to_insert_t create_zero(
+    laddr_t addr, extent_len_t len) {
+    return extent_to_insert_t(addr, len);
+  }
+
+private:
+  extent_to_insert_t(laddr_t addr, extent_len_t len,
+    std::optional<bufferlist> bl)
+    :type(type_t::DATA), addr(addr), len(len), bl(bl) {}
+
+  extent_to_insert_t(laddr_t addr, extent_len_t len)
+    :type(type_t::ZERO), addr(addr), len(len) {}
+};
+using extent_to_insert_list_t = std::list<extent_to_insert_t>;
+
+// Encapsulates extents to be retired in do_removals.
+using extent_to_remove_list_t = std::list<LBAPinRef>;
+
+struct overwrite_ops_t {
+  extent_to_remap_list_t to_remap = {};
+  extent_to_insert_list_t to_insert = {};
+  extent_to_remove_list_t to_remove;
+};
+
+// prepare to_remap, to_retire, to_insert list
+overwrite_ops_t prepare_ops_list(
+  lba_pin_list_t &pins,
+  extent_to_write_list_t &to_write) {
+  assert(to_write.size() != 0 && pins.size() != 0);
+  overwrite_ops_t ops;
+  ops.to_remove.swap(pins);
+  auto& front = to_write.front();
+  auto& back = to_write.back();
+
+  // overwrite happen in one original extent.
+  if (ops.to_remove.size() == 1 &&
+    (front.is_existing() || back.is_existing())) {
+    assert(front.is_existing() ? bool(front.pin) : !bool(front.pin)); 
+    assert(back.is_existing() ? bool(back.pin) : !bool(back.pin)); 
+    assert(to_write.size() == 2 || to_write.size() == 3);
+    if (front.is_existing()) {
+      auto& overwrite = *(++to_write.begin());
+      ops.to_remap.push_back(extent_to_remap_t::create_overwrite(
+        std::move(front.pin),
+        overwrite.addr - front.pin->get_key(),
+        overwrite.len,
+        overwrite.to_write));
+    } else {
+      auto& overwrite = *(to_write.begin());
+      ops.to_remap.push_back(extent_to_remap_t::create_overwrite(
+        std::move(back.pin),
+        overwrite.addr - back.pin->get_key(),
```

Comment:
One question here. Is the following statement always true? `extent_to_write_list_t.addr > back.pin->get_key() ` If so, we might add assert() because laddr is unit64_t.

Reply:
It's always true. I added an assert here.